File: domain/identity.ts
/**
 * domain/identity.ts
 *
 * Identity is the central aggregation point of the FOOL Platform domain
 * model, per the Identity-Centric Architecture principle. It resolves a set
 * of observable identifiers (email, phone, username, IP, wallet address,
 * etc.) into a single subject of interest, and is the anchor that Entities,
 * Evidence, and Findings reference directly or indirectly.
 *
 * Mirrors contracts/domain/identity.schema.json field-for-field.
 *
 * Domain Purity: standard runtime and sibling domain modules only.
 */

import {
  Status,
  createProvenance,
  freezeRefs,
  freezeTags,
  newId,
  now,
  type Id,
  type Metadata,
  type Provenance,
  type Reference,
  type Timestamp,
  EMPTY_METADATA,
} from "./common.js";
import { ClassificationLevel } from "./classification_level.js";
import { ConfidenceScore } from "./confidence_score.js";

export enum IdentifierType {
  Email = "email",
  Phone = "phone",
  Username = "username",
  Domain = "domain",
  IpAddress = "ip_address",
  Hash = "hash",
  WalletAddress = "wallet_address",
  Document = "document",
  Image = "image",
  PersonName = "person_name",
  Other = "other",
}

export interface IdentifierInit {
  readonly identifierType: IdentifierType;
  readonly value: string;
  readonly confidence: ConfidenceScore;
  readonly sourceRefs?: ReadonlyArray<Reference>;
}

export interface Identifier extends IdentifierInit {
  readonly sourceRefs: ReadonlyArray<Reference>;
}

export interface IdentityInit {
  readonly identifiers: ReadonlyArray<IdentifierInit>;
  readonly displayLabel?: string;
  readonly status?: Status;
  readonly classification?: ClassificationLevel;
  readonly entityRefs?: ReadonlyArray<Reference>;
  readonly evidenceRefs?: ReadonlyArray<Reference>;
  readonly confidence?: ConfidenceScore;
  readonly provenance: Provenance;
  readonly tags?: ReadonlyArray<string>;
  readonly metadata?: Metadata;
}

/**
 * Identity is an immutable aggregate of resolved identifiers. Every mutation
 * (adding an identifier, linking an entity) returns a new Identity rather
 * than modifying the existing instance, preserving referential history.
 */
export class Identity {
  readonly id: Id;
  readonly version: string;
  readonly createdAt: Timestamp;
  readonly updatedAt: Timestamp;
  readonly status: Status;
  readonly classification: ClassificationLevel;
  readonly displayLabel?: string;
  readonly identifiers: ReadonlyArray<Identifier>;
  readonly entityRefs: ReadonlyArray<Reference>;
  readonly evidenceRefs: ReadonlyArray<Reference>;
  readonly confidence?: ConfidenceScore;
  readonly provenance: Provenance;
  readonly tags: ReadonlyArray<string>;
  readonly metadata: Metadata;

  private constructor(
    id: Id,
    version: string,
    createdAt: Timestamp,
    updatedAt: Timestamp,
    init: IdentityInit
  ) {
    if (init.identifiers.length === 0) {
      throw new Error("Identity must be created with at least one identifier.");
    }
    this.id = id;
    this.version = version;
    this.createdAt = createdAt;
    this.updatedAt = updatedAt;
    this.status = init.status ?? Status.Active;
    this.classification = init.classification ?? ClassificationLevel.Internal;
    this.displayLabel = init.displayLabel;
    this.identifiers = Object.freeze(
      init.identifiers.map((i) => ({ ...i, sourceRefs: freezeRefs(i.sourceRefs) }))
    );
    this.entityRefs = freezeRefs(init.entityRefs);
    this.evidenceRefs = freezeRefs(init.evidenceRefs);
    this.confidence = init.confidence;
    this.provenance = init.provenance;
    this.tags = freezeTags(init.tags);
    this.metadata = init.metadata ?? EMPTY_METADATA;
    Object.freeze(this);
  }

  static create(init: IdentityInit): Identity {
    const timestamp = now();
    return new Identity(newId(), "1.0.0", timestamp, timestamp, init);
  }

  /** Returns a new Identity with the given identifier appended. */
  withIdentifier(identifier: IdentifierInit): Identity {
    return new Identity(this.id, this.version, this.createdAt, now(), {
      identifiers: [...this.identifiers, identifier],
      displayLabel: this.displayLabel,
      status: this.status,
      classification: this.classification,
      entityRefs: this.entityRefs,
      evidenceRefs: this.evidenceRefs,
      confidence: this.confidence,
      provenance: this.provenance,
      tags: this.tags,
      metadata: this.metadata,
    });
  }

  /** Returns a new Identity linked to the given Entity reference. */
  withEntityRef(entityRef: Reference): Identity {
    if (this.entityRefs.some((ref) => ref.refId === entityRef.refId)) {
      return this;
    }
    return new Identity(this.id, this.version, this.createdAt, now(), {
      identifiers: this.identifiers,
      displayLabel: this.displayLabel,
      status: this.status,
      classification: this.classification,
      entityRefs: [...this.entityRefs, entityRef],
      evidenceRefs: this.evidenceRefs,
      confidence: this.confidence,
      provenance: this.provenance,
      tags: this.tags,
      metadata: this.metadata,
    });
  }

  hasIdentifierValue(value: string): boolean {
    return this.identifiers.some((identifier) => identifier.value === value);
  }
}

/** Convenience factory mirroring createProvenance for callers assembling an Identity inline. */
export function openIdentity(
  origin: string,
  identifiers: ReadonlyArray<IdentifierInit>,
  options: Omit<IdentityInit, "identifiers" | "provenance"> = {}
): Identity {
  return Identity.create({ ...options, identifiers, provenance: createProvenance(origin) });
}

File: domain/identity.py
"""
domain/identity.py

Identity is the central aggregation point of the FOOL Platform domain model.
It resolves a set of observable identifiers into a single subject of interest.
Mirrors contracts/domain/identity.schema.json field-for-field.
"""
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, FrozenSet, Sequence

from .classification_level import ClassificationLevel, ClassificationLevelValue
from .confidence_score import ConfidenceScore, ConfidenceLevel, ConfidenceMethods
from .common import (



    DomainInvariantError,
    EMPTY_METADATA,
    Metadata,
    Provenance,
    Reference,
    Status,
    freeze_refs,
    freeze_tags,
    new_id,
    utc_now,
    create_provenance,
)


class IdentifierType(str, Enum):
    """Types of observable identifiers."""
    EMAIL = "email"
    PHONE = "phone"
    USERNAME = "username"
    DOMAIN = "domain"
    IP_ADDRESS = "ip_address"
    HASH = "hash"
    WALLET_ADDRESS = "wallet_address"
    DOCUMENT = "document"
    IMAGE = "image"
    PERSON_NAME = "person_name"
    OTHER = "other"


@dataclass(frozen=True)
class Identifier:
    """An observable identifier resolved to belong to an Identity."""
    identifier_type: IdentifierType
    value: str
    confidence: ConfidenceScore
    source_refs: FrozenSet[Reference] = field(default_factory=frozenset)


@dataclass(frozen=True)
class Identity:
    """
    Identity is an immutable aggregate of resolved identifiers.
    
    Identity is the central aggregation point of the FOOL Platform domain model.
    Every identifier, entity, evidence item, and finding either references an
    Identity directly or reaches one transitively.
    """
    id: str
    version: str
    created_at: str
    updated_at: str
    status: Status
    classification: ClassificationLevel
    provenance: Provenance
    identifiers: tuple[Identifier, ...]
    display_label: str | None = None
    entity_refs: FrozenSet[Reference] = field(default_factory=frozenset)
    evidence_refs: FrozenSet[Reference] = field(default_factory=frozenset)
    confidence: ConfidenceScore | None = None
    tags: FrozenSet[str] = field(default_factory=frozenset)
    metadata: Metadata = field(default_factory=lambda: dict(EMPTY_METADATA))

    def __post_init__(self) -> None:
        if len(self.identifiers) == 0:
            raise DomainInvariantError(
                "Identity must be created with at least one identifier."
            )

    @classmethod
    def create(
        cls,
        identifiers: Sequence[Identifier],
        provenance: Provenance,
        display_label: str | None = None,
        status: Status = Status.ACTIVE,
        classification: ClassificationLevelValue = ClassificationLevelValue.INTERNAL,
        entity_refs: Sequence[Reference] | None = None,
        evidence_refs: Sequence[Reference] | None = None,
        confidence: ConfidenceScore | None = None,
        tags: Sequence[str] | None = None,
        metadata: Metadata | None = None,
    ) -> "Identity":
        """
        Create a new Identity.
        
        Args:
            identifiers: At least one identifier to resolve to this identity
            provenance: Origin and lineage of this identity
            display_label: Optional human-readable label
            status: Lifecycle status (default: active)
            classification: Classification level (default: internal)
            entity_refs: Optional references to linked entities
            evidence_refs: Optional references to supporting evidence
            confidence: Optional overall confidence score
            tags: Optional tags for categorization
            metadata: Optional extension attributes
        """
        if len(identifiers) == 0:
            raise DomainInvariantError(
                "Identity must be created with at least one identifier."
            )
        
        timestamp = utc_now()
        return cls(
            id=new_id(),
            version="1.0.0",
            created_at=timestamp,
            updated_at=timestamp,
            status=status,
            classification=ClassificationLevel(classification),
            provenance=provenance,
            identifiers=tuple(identifiers),
            display_label=display_label,
            entity_refs=frozenset(entity_refs) if entity_refs else frozenset(),
            evidence_refs=frozenset(evidence_refs) if evidence_refs else frozenset(),
            confidence=confidence,
            tags=frozenset(tags) if tags else frozenset(),
            metadata=dict(metadata) if metadata else dict(EMPTY_METADATA),
        )

    def with_identifier(self, identifier: Identifier) -> "Identity":
        """Return a new Identity with the given identifier appended."""
        return Identity(
            id=self.id,
            version=self.version,
            created_at=self.created_at,
            updated_at=utc_now(),
            status=self.status,
            classification=self.classification,
            provenance=self.provenance,
            identifiers=(*self.identifiers, identifier),
            display_label=self.display_label,
            entity_refs=self.entity_refs,
            evidence_refs=self.evidence_refs,
            confidence=self.confidence,
            tags=self.tags,
            metadata=self.metadata,
        )

    def with_entity_ref(self, entity_ref: Reference) -> "Identity":
        """Return a new Identity linked to the given Entity reference."""
        if entity_ref in self.entity_refs:
            return self
        return Identity(
            id=self.id,
            version=self.version,
            created_at=self.created_at,
            updated_at=utc_now(),
            status=self.status,
            classification=self.classification,
            provenance=self.provenance,
            identifiers=self.identifiers,
            display_label=self.display_label,
            entity_refs=self.entity_refs | {entity_ref},
            evidence_refs=self.evidence_refs,
            confidence=self.confidence,
            tags=self.tags,
            metadata=self.metadata,
        )

    def has_identifier_value(self, value: str) -> bool:
        """Check if this identity has an identifier with the given value."""
        return any(identifier.value == value for identifier in self.identifiers)


def open_identity(
    origin: str,
    identifiers: Sequence[Identifier],
    display_label: str | None = None,
    status: Status = Status.ACTIVE,
    classification: ClassificationLevelValue = ClassificationLevelValue.INTERNAL,
    entity_refs: Sequence[Reference] | None = None,
    evidence_refs: Sequence[Reference] | None = None,
    confidence: ConfidenceScore | None = None,
    tags: Sequence[str] | None = None,
    metadata: Metadata | None = None,
) -> Identity:
    """
    Convenience factory to open a new identity with provenance from an origin.
    """
    return Identity.create(
        identifiers=identifiers,
        provenance=create_provenance(origin),
        display_label=display_label,
        status=status,
        classification=classification,
        entity_refs=entity_refs,
        evidence_refs=evidence_refs,
        confidence=confidence,
        tags=tags,
        metadata=metadata,
    )


__all__ = [
    "Identifier",
    "IdentifierType",
    "Identity",
    "open_identity",
]
